src/tree_builder/actions.rs

```diff
@@ -775,10 +743,40 @@ impl<Handle, Sink> TreeBuilderActions<Handle>
         // FIXME: application cache selection algorithm
     }
 
+    // https://html.spec.whatwg.org/multipage/#create-an-element-for-the-token
     fn insert_element(&mut self, push: PushFlag, ns: Namespace, name: LocalName, attrs: Vec<Attribute>)
             -> Handle {
-        let elem = self.sink.create_element(QualName::new(ns, name), attrs);
-        self.insert_appropriately(AppendNode(elem.clone()), None);
+        declare_tag_set!(form_associatable =
+            "button" "fieldset" "input" "object"
+            "output" "select" "textarea" "img");
+
+        declare_tag_set!(listed = [form_associatable] - "img");
+
+        // Step 7.
+        let qname = QualName::new(ns, name);
+        let elem = self.sink.create_element(qname.clone(), attrs.clone());
+
+        let insertion_point = self.appropriate_place_for_insertion(None);
+        let tree_node = match insertion_point {
+            LastChild(ref p) |
+            BeforeSibling(ref p) => p.clone()
+        };
+
+        // Step 12.
+        // TODO: Handle template element case
```

Comment:
Yes, you're right. I forgot about it. Implemented it now, but I need to be sure that it's right :) https://html.spec.whatwg.org/multipage/syntax.html#create-an-element-for-the-token step 12 says:
> ...and there is no template element on the stack of open elements...

I hope I understand it right.